rust/template/d3log/src/lib.rs

```diff
@@ -0,0 +1,228 @@
+pub mod error;
+mod forwarder;
+pub mod record_batch;
+
+use core::fmt;
+use differential_datalog::{ddval::DDValue, record::*, D3logLocationId};
+use std::borrow::Cow;
+use std::fmt::Display;
+use std::sync::{Arc, Mutex};
+use tokio::runtime::Runtime;
+
+use crate::{
+    broadcast::{Broadcast, PubSub},
+    ddvalue_batch::DDValueBatch,
+    dispatch::Dispatch,
+    error::Error,
+    forwarder::Forwarder,
+    record_batch::RecordBatch,
+};
+
+pub type Node = D3logLocationId;
+
+pub trait EvaluatorTrait {
+    fn ddvalue_from_record(&self, id: String, r: Record) -> Result<DDValue, Error>;
+    fn eval(&self, input: Batch) -> Result<Batch, Error>;
+    fn id_from_relation_name(&self, s: String) -> Result<usize, Error>;
+    fn localize(&self, rel: usize, v: DDValue) -> Option<(Node, usize, DDValue)>;
+    fn now(&self) -> u64;
+    fn myself(&self) -> Node;
+    fn error(&self, text: Record, line: Record, filename: Record, functionname: Record);
+    fn record_from_ddvalue(&self, d: DDValue) -> Result<Record, Error>;
+    fn relation_name_from_id(&self, id: usize) -> Result<String, Error>;
+
+    // these is ddvalue/relationid specific
+    fn serialize_batch(&self, b: DDValueBatch) -> Result<Vec<u8>, Error>;
+    fn deserialize_batch(&self, s: Vec<u8>) -> Result<DDValueBatch, Error>;
+}
+
+pub type Evaluator = Arc<(dyn EvaluatorTrait + Send + Sync)>;
+
+#[derive(Clone)]
+pub enum Batch {
+    Value(DDValueBatch),
+    Rec(RecordBatch),
+}
+
+impl Display for Batch {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        writeln!(f, "Batch [").unwrap();
+        match self {
+            Batch::Value(b) => b.fmt(f),
+            Batch::Rec(b) => b.fmt(f),
+        }
+        .unwrap();
+        writeln!(f, "\n]\n")
+    }
+}
+
+pub trait Transport {
+    // since most of these errors are async, we're adopting a general
+    // policy for the moment of making all errors async and reported out
+    // of band.
+
+    // should really be type parametric shouldn't it?
+    fn send(&self, b: Batch);
+}
+
+pub type Port = Arc<(dyn Transport + Send + Sync)>;
+
+struct AccumulatePort {
+    eval: Evaluator,
+    b: Arc<Mutex<DDValueBatch>>,
+}
+
+impl Transport for AccumulatePort {
+    fn send(&self, b: Batch) {
+        for (r, f, w) in &DDValueBatch::from(&(*self.eval), b).expect("iterator") {
+            self.b.lock().expect("lock").insert(r, f, w);
+        }
+    }
+}
+
+use std::collections::VecDeque;
+
+struct EvalPort {
+    eval: Evaluator,
+    forwarder: Port,
+    dispatch: Port,
+    queue: Arc<Mutex<VecDeque<Batch>>>,
+}
+
+impl Transport for EvalPort {
+    fn send(&self, b: Batch) {
+        println!(
+            "ep {} {}",
+            self.eval.clone().myself(),
+            RecordBatch::from(self.eval.clone(), b.clone())
+        );
+        self.dispatch.send(b.clone());
+
+        {
+            self.queue.lock().expect("lock").push_back(b.clone());
+        }
+
+        loop {
+            let b = {
```

Comment:
this scheduling has already shows itself to be misguided when running with multiple threads. will update.

Reply:
unsurprisingly, using a proper mpsc here works just fine...will fix here